ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFDateFormat.java

```diff
@@ -111,17 +123,18 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
     // the function should support both short date and full timestamp format
     // time part of the timestamp should not be skipped
     Timestamp ts = getTimestampValue(arguments, 0, tsConverters);
+
     if (ts == null) {
       Date d = getDateValue(arguments, 0, dtInputTypes, dtConverters);
       if (d == null) {
         return null;
       }
       ts = Timestamp.ofEpochMilli(d.toEpochMilli(id), id);
     }
-
-
-    date.setTime(ts.toEpochMilli(id));
-    String res = formatter.format(date);
+    Timestamp ts2 = TimestampTZUtil.convertTimestampToZone(ts, timeZone, ZoneId.of("UTC"));
+    Instant instant = Instant.ofEpochSecond(ts2.toEpochSecond(), ts2.getNanos());
+    ZonedDateTime zonedDateTime = ZonedDateTime.ofInstant(instant, ZoneOffset.UTC);
+    String res = formatter.format(zonedDateTime);
```

Comment:
Instead use ts.format(formatter)

Reply:
Wrong results without conversion

Reply:
The timezone gets converted for some specific locale's in case we don't do this conversion:
PDT -> PT
CST -> CT

Reply:
You can create a new function like formatWithTimeZone(formatter, ZoneId) and add timeZone value to localDateTime.